### PR TITLE
Get Set , Load User Cache

### LINKED ISSUE #102: Misc.9 Client Side User Cache Services
## User Story
Implement client side user cache service to reduce the number of calls to the database

## Sub-Tasks
- [ ] create user services that uses local/session storage to cache and retrieve data
- [ ] create user services that first call the local storage service. On cache miss, use the Firebase service version to retrieve the data. Posts/Puts should save data in both locally and remotely
- [ ] Update providers in the app config file if necessary

### PR DESCRIPTION
**Pull request by JeremyLin003**
fixes #102 

When getting user or study set for the first time, stores the object as a string (JSON) in session storage
I haven't tested the ones that call firebase
uid is also stored under key "userId" to allow for easy access of user info without needing to pass uid between every page

### CODE REVIEW

File: src/app/services/user-info-firebase.service.ts [hunk 1]
```diff
@@ -4,29 +4,44 @@ import {
...
+    } else {
+      sessionStorage.clear();
+      let userVar:UserData;
+      let userObservable =  defer(() =>
+        from(getDoc(doc(this.firestore, 'users', id)) as Promise<DocumentSnapshot>))
+        .pipe(switchMap((docSnap: DocumentSnapshot) =>
+          iif(() => !docSnap.exists(),
+            this.saveUser(new UserData(id, [], [])).pipe(map((dbUser) => new UserData(dbUser, [], []))),
+            of(docSnap.data() as UserModel).pipe(map((dbUser: UserModel) => UserData.copyUser(dbUser))
+          ))
```
Comment (ashkjones) [inline src/app/services/user-info-firebase.service.ts:32]: This doesn't appear to be saving the owned and viewed study sets of the user. 

Console log showing that the correct user data was fetched from the firestore, but empty arrays being created
image

On user cache hit after page reload, we get an empty page
image

**Review by ashkjones — CHANGES_REQUESTED** [review]
Please fix how the user is saved locally when using the firestore. I can help you set up the local emulator if you need me to.

Great work though!!

File: src/app/services/user-info-fake.service.ts [hunk 2]
```diff
@@ -1,66 +1,76 @@
...
-        "viewed": "2021-10-16T19:32:33.420Z"
-      },
-      {
-        "setId": "llll",
-        "viewed": "2022-03-30T19:32:33.420Z"
-      }
-    ];
-    return of(new UserData("user1", setList, setList));
+
+    if (id === sessionStorage.getItem("userId")) {
```
Comment (B-Yamawaki) [inline src/app/services/user-info-fake.service.ts:12]: Could you change this to checking if the actual user has been created. `if (sessionStorage.getItem(id))` It doesn't change too much, but it helps make things easier for the sign in implementation. And the same in the firebase service.

Comment (B-Yamawaki) [reply src/app/services/user-info-fake.service.ts:12]: You also don't have to set userId, as I can do that upon sign in.

**Pull request merged**